iree/hal/local/executable_library.h

```diff
@@ -194,6 +206,68 @@ typedef struct iree_hal_executable_import_table_v0_t {
   const char* const* symbols;
 } iree_hal_executable_import_table_v0_t;
 
+// Maximum number of data fields in iree_hal_processor_v0_t.
+// Executables must only reference the specified field count and this is allowed
+// to grow at any time up to 32 uint64_t values (256B).
+#define IREE_HAL_PROCESSOR_DATA_CAPACITY_V0 4
+
+// Architecture-specific CPU information available to executables.
+// This encodes a high-level identifier and zero or more fields of opaque data.
+// It's encoded as a variable-length array of uint64_t values and if no
+// additional data is provided is bitwise equivalent to a single 64-bit value.
+// The intent is that this structure can be put in .rodata when there are no
+// runtime features that need to be queried.
+typedef struct iree_hal_processor_v0_t {
+  // Simple identifier or set of flags that can be used to quickly test for
+  // major architectural features without the need to walk the data table. The
+  // format of the field is architecture-specific as by construction no value
+  // will ever be used in a compiled binary from another architecture. This
+  // allows us to simplify this interface as we can't for example load the same
+  // executable library for both aarch64 on riscv32 and don't need to normalize
+  // any of the fields across them both.
+  //
+  // The contents are opaque here as to support out-of-tree architectures. The
+  // runtime code deriving the identifier/flags and providing it here is losely
+  // coupled with the compiler code emitting checks based on the identifier and
+  // only those two places ever need to change.
+  uint32_t flags;
+
+  // Bitfield indicating which data fields are valid.
+  // Example: ...0101: fields data[0] and data[2] are valid.
+  uint32_t data_valid;
```

Comment:
Can you elaborate on why this `data_valid` bitfield exists? Would `data_size` (used size vs maximum data capacity) be simpler?

I'm coming at this from the perspective of larger or more complex data structure than what 4 64 bit values implies by default (serialized protobuf, struct with some floats/doubles/strings). For example, I'd find it easier to use a single size instead of needing to check/count that the bitfield is `11111111111111111111111100000000` (size 24) vs `11111111111110111111111100000000` (almost size 24, one invalid entry in the middle).

Can you give an example of the sort of "fixed-length data structure, a series of hardware registers, or key-value pairs" that would use this bitfield?

edit: I see the associated `loadProcessorData` function in `iree/compiler/Codegen/LLVMCPU/ConvertToLLVM.cpp`. I can see how this would be easier to use from compiler generated code, but I'm still not fully grasping the concrete details.

Reply:
Good q - it was a hunch that we'd have sparse fields, but that may not be worth it.

It really depends on what we end up putting in here and how we want to version it. The important thing we have to support is the unavailability of information either due to the runtime being older than the executable being loaded (so some fields won't be present) or the information not being available (not present on the current architecture, not able to be queried due to sandboxing/permissions/etc).

So if for example we had on aarch64 a representation like:
```
typedef union {
    struct {
      uint32_t has_dotprod : 1;
      uint32_t has_i8mm : 1;
    };
    uint32_t feature_bits;
} aarch64_features_t;
typedef struct {
  uint64_t AA64ISAR0_EL1;
  uint64_t AA64ISAR1_EL1;
} aarch64_info_t;
```
The feature bits would be the things packed in to the fixed data structure and something always available. A 0 in the feature field indicates the runtime doesn't know about the feature but does not indicate the feature is unavailable: it may just be the runtime was compiled before the specific architecture revision it's running on was released (or support was added). The register fields, though, are both optional: the data_valid bits being 0 indicates the registers aren't available, while the bits within the register fields being 0 indicate the feature isn't available. Think of data_valid like Optional<> - it's a way to tell whether a 0 indicates the data isn't available or the feature isn't available. Where the sparsity becomes interesting is that we can't ever reorder the fields once we go stable; if we add a 3rd field then:
```mlir
typedef struct {
  uint64_t AA64ISAR0_EL1;
  uint64_t AA64ISAR1_EL1;
  uint64_t more_feature_bits;
} aarch64_info_t;
```
we need a way to know that field[2] is valid even if field[0] & field[1] aren't available. A size here wouldn't help as that would imply the fields 0-2 are valid.

That said, if we make the decision that 0's are meaningful field values then we could drop the data_valid bits - in the case above with aarch64 registers this works out ok - but I wasn't sure that would hold for all possible architectures or all fields we may want to add. It's simpler, though, and may be worth it just for that even if other things get more complex.

Typing this up may have convinced me to rip it out and just make it all flat fields :)

Reply:
Simplified to just a flag list of fields. 512 bits should be enough for anyone and architectures can implement bitfields like data_valid themselves if needed.